File: packages/EarthSciSerialization.jl/src/expression.jl
"""
Expression substitution and structural operations module.

This module provides functions for working with ESM format expressions:
- Variable substitution with scoped reference support
- Free variable analysis
- Expression simplification through constant folding

All operations are non-mutating and return new Expr objects.

Numerical evaluation lives in `tree_walk.jl` (`evaluate_expr` /
`build_evaluator`) — the official ESS Julia evaluator — so this module
hosts no parallel dispatch table. `simplify`'s constant-folding step
delegates to `evaluate_expr` so adding an op to the tree-walk evaluator
transparently extends the folder.
"""

# ========================================
# 1. Variable Substitution
# ========================================

"""
    substitute(expr::Expr, bindings::Dict{String,Expr})::Expr

Recursively replace variables in an expression with provided bindings.
Supports scoped reference resolution - if a variable is not found in bindings,
it remains unchanged. Returns a new Expr object (non-mutating).

# Arguments
- `expr`: The expression to perform substitution on
- `bindings`: Dictionary mapping variable names to replacement expressions

# Examples
```julia
# Simple substitution
x = VarExpr("x")
y = VarExpr("y")
sum_expr = OpExpr("+", [x, y])
bindings = Dict("x" => NumExpr(2.0))
result = substitute(sum_expr, bindings)  # OpExpr("+", [NumExpr(2.0), VarExpr("y")])

# Nested substitution
nested = OpExpr("*", [OpExpr("+", [x, NumExpr(1.0)]), y])
result = substitute(nested, bindings)  # OpExpr("*", [OpExpr("+", [NumExpr(2.0), NumExpr(1.0)]), VarExpr("y")])
```
"""
function substitute(expr::NumExpr, bindings::Dict{String,Expr})::Expr
    return expr  # Numeric literals are unchanged
end

function substitute(expr::IntExpr, bindings::Dict{String,Expr})::Expr
    return expr  # Integer literals are unchanged
end

function substitute(expr::VarExpr, bindings::Dict{String,Expr})::Expr
    return get(bindings, expr.name, expr)  # Replace if bound, otherwise keep original
end

function substitute(expr::OpExpr, bindings::Dict{String,Expr})::Expr
    # Substitute into EVERY sub-expression the node carries — not just `args` —
    # so substitution is complete inside aggregate/arrayop bodies, filter
    # predicates, integral bounds, makearray values, and table_lookup axis
    # inputs. `reconstruct` then preserves all other fields (semiring, ranges,
    # output_idx, table, manifold, id, join, …) that earlier hand-listed
    # rebuilds silently dropped. Bound locals (index vars, `int_var`) are short
    # local symbols never present in `bindings` (whose keys are namespaced
    # globals / parameter names), so recursing cannot capture them.
    sub(x) = x === nothing ? nothing : substitute(x, bindings)
    new_args = Expr[substitute(arg, bindings) for arg in expr.args]
    new_values = expr.values === nothing ? nothing :
        Expr[substitute(v, bindings) for v in expr.values]
    new_table_axes = expr.table_axes === nothing ? nothing :
        Dict{String,Expr}(k => substitute(v, bindings) for (k, v) in expr.table_axes)
    return reconstruct(expr;
        args = new_args,
        lower = sub(expr.lower), upper = sub(expr.upper),
        expr_body = sub(expr.expr_body), filter = sub(expr.filter),
        values = new_values, table_axes = new_table_axes)
end

# ========================================
# 2. Free Variable Analysis
# ========================================

"""
    free_variables(expr::Expr)::Set{String}

Extract all free (unbound) variable names from an expression.
Returns a set of variable names that appear in the expression.

# Examples
```julia
x = VarExpr("x")
y = VarExpr("y")
sum_expr = OpExpr("+", [x, y])
vars = free_variables(sum_expr)  # Set(["x", "y"])

nested = OpExpr("*", [OpExpr("+", [x, NumExpr(1.0)]), y])
vars = free_variables(nested)  # Set(["x", "y"])
```
"""
function free_variables(expr::NumExpr)::Set{String}
    return Set{String}()  # No variables in numeric literals
end

function free_variables(expr::IntExpr)::Set{String}
    return Set{String}()  # No variables in integer literals
end

function free_variables(expr::VarExpr)::Set{String}
    return Set([expr.name])  # Single variable
end

function free_variables(expr::OpExpr)::Set{String}
    # Union of free variables from all arguments
    result = Set{String}()
    for arg in expr.args
        union!(result, free_variables(arg))
    end

    # Add variables from wrt field if present
    if expr.wrt !== nothing
        push!(result, expr.wrt)
    end

    return result
end

# ========================================
# 3. Variable Containment Check
# ========================================

"""
    contains(expr::Expr, var::String)::Bool

Check if an expression contains a specific variable name.
Returns true if the variable appears anywhere in the expression.

# Examples
```julia
x = VarExpr("x")
y = VarExpr("y")
sum_expr = OpExpr("+", [x, y])
contains(sum_expr, "x")  # true
contains(sum_expr, "z")  # false
```
"""
function contains(expr::IntExpr, var::String)::Bool
    return false  # Integer literals don't contain variables
end

function contains(expr::NumExpr, var::String)::Bool
    return false  # Numeric literals don't contain variables
end

function contains(expr::VarExpr, var::String)::Bool
    return expr.name == var
end

function contains(expr::OpExpr, var::String)::Bool
    # Check if any argument contains the variable
    for arg in expr.args
        if contains(arg, var)
            return true
        end
    end

    # Check wrt field
    if expr.wrt !== nothing && expr.wrt == var
        return true
    end

    return false
end

# ========================================
# 4. Evaluation error type
# ========================================

"""
    UnboundVariableError

Raised by [`evaluate_expr`](@ref) (the tree-walk evaluator entry point)
when an expression references a variable name that is not in the supplied
bindings. Defined here so it is in scope for callers that catch the
"binding not yet resolved" signal during iterated observed-variable
fixed-point passes.
"""
struct UnboundVariableError <: Exception
    variable_name::String
    message::String
end

Base.show(io::IO, e::UnboundVariableError) = print(io, "UnboundVariableError: $(e.message)")

# ========================================
# 5. Expression Simplification
# ========================================

"""
    simplify(expr::Expr)::Expr

Perform constant folding and algebraic simplification on an expression.
Returns a new simplified Expr object (non-mutating).

# Simplification Rules
- Constant folding: `2 + 3` → `5`
- Additive identity: `x + 0` → `x`, `0 + x` → `x`
- Multiplicative identity: `x * 1` → `x`, `1 * x` → `x`
- Multiplicative zero: `x * 0` → `0`, `0 * x` → `0`
- Exponentiation: `x^0` → `1`, `x^1` → `x`

# Examples
```julia
# Constant folding
expr = OpExpr("+", [NumExpr(2.0), NumExpr(3.0)])
result = simplify(expr)  # NumExpr(5.0)

# Identity elimination
expr = OpExpr("*", [VarExpr("x"), NumExpr(1.0)])
result = simplify(expr)  # VarExpr("x")
```
"""
function simplify(expr::NumExpr)::Expr
    return expr  # Already simplified
end

function simplify(expr::IntExpr)::Expr
    return expr  # Already simplified
end

function simplify(expr::VarExpr)::Expr
    return expr  # Already simplified
end

"""
    is_literal(expr::Expr)::Bool

True iff `expr` is a numeric literal (either integer or float node).
"""
is_literal(expr::Expr)::Bool = isa(expr, NumExpr) || isa(expr, IntExpr)

"""
    literal_value(expr::Expr)

Return the numeric value of a literal node, as `Float64`. Throws on non-literals.
"""
literal_value(expr::NumExpr) = expr.value
literal_value(expr::IntExpr) = Float64(expr.value)

function simplify(expr::OpExpr)::Expr
    # First recursively simplify all arguments
    simplified_args = Expr[simplify(arg) for arg in expr.args]
    lower_simplified = expr.lower === nothing ? nothing : simplify(expr.lower)
    upper_simplified = expr.upper === nothing ? nothing : simplify(expr.upper)

    op = expr.op

    # Try constant folding first - if all arguments are numeric, evaluate.
    # Per RFC §5.4.1, promotion happens only in evaluate, not simplify;
    # so when mixed int/float inputs fold, the result is a float literal.
    # When all inputs are integer, preserve integer result (for ops whose
    # integer result is representable — fall back to float on non-integer).
    # Folding is delegated to the official tree-walk evaluator
    # (`evaluate_expr`) so the simplifier shares the runner's dispatch
    # table and there is no parallel operator switch in this module.
    if all(is_literal, simplified_args)
        try
            result_value = evaluate_expr(
                reconstruct(expr; args=simplified_args, lower=lower_simplified, upper=upper_simplified),
                Dict{String,Float64}())
            all_int = all(arg -> isa(arg, IntExpr), simplified_args)
            if all_int && isfinite(result_value) && result_value == trunc(result_value) &&
               abs(result_value) <= Float64(typemax(Int64))
                return IntExpr(Int64(result_value))
            end
            return NumExpr(result_value)
        catch
            # If evaluation fails, continue with algebraic simplification
        end
    end

    # Helper: true when arg is a numeric literal equal to v (compared by value).
    is_lit_val(arg, v) = (isa(arg, NumExpr) && arg.value == v) ||
                         (isa(arg, IntExpr) && Float64(arg.value) == v)

    # Algebraic simplification rules
    if op == "+"
        # Remove zeros: x + 0 = x, 0 + x = x
        non_zero_args = filter(arg -> !is_lit_val(arg, 0.0), simplified_args)
        if length(non_zero_args) == 0
            return NumExpr(0.0)
        elseif length(non_zero_args) == 1
            return non_zero_args[1]
        else
            return reconstruct(expr; args=Expr[non_zero_args...], lower=lower_simplified, upper=upper_simplified)
        end

    elseif op == "*"
        # Check for zeros: x * 0 = 0, 0 * x = 0
        for arg in simplified_args
            if is_lit_val(arg, 0.0)
                return NumExpr(0.0)
            end
        end

        # Remove ones: x * 1 = x, 1 * x = x
        non_one_args = filter(arg -> !is_lit_val(arg, 1.0), simplified_args)
        if length(non_one_args) == 0
            return NumExpr(1.0)
        elseif length(non_one_args) == 1
            return non_one_args[1]
        else
            return reconstruct(expr; args=Expr[non_one_args...], lower=lower_simplified, upper=upper_simplified)
        end

    elseif op == "^" && length(simplified_args) == 2
        base = simplified_args[1]
        exponent = simplified_args[2]

        # x^0 = 1
        if is_lit_val(exponent, 0.0)
            return NumExpr(1.0)
        end

        # x^1 = x
        if is_lit_val(exponent, 1.0)
            return base
        end

        # 0^x = 0 (for x > 0)
        if is_lit_val(base, 0.0) && is_literal(exponent) && literal_value(exponent) > 0.0
            return NumExpr(0.0)
        end

        # 1^x = 1
        if is_lit_val(base, 1.0)
            return NumExpr(1.0)
        end

        return reconstruct(expr; args=simplified_args, lower=lower_simplified, upper=upper_simplified)

    elseif op == "-" && length(simplified_args) == 2
        # x - 0 = x
        if is_lit_val(simplified_args[2], 0.0)
            return simplified_args[1]
        end

        return reconstruct(expr; args=simplified_args, lower=lower_simplified, upper=upper_simplified)

    elseif op == "/" && length(simplified_args) == 2
        # x / 1 = x
        if is_lit_val(simplified_args[2], 1.0)
            return simplified_args[1]
        end

        # 0 / x = 0 (for x != 0)
        if isa(simplified_args[1], NumExpr) && simplified_args[1].value == 0.0
            return NumExpr(0.0)
        end

        return reconstruct(expr; args=simplified_args, lower=lower_simplified, upper=upper_simplified)
    end

    # If no simplification rules apply, return the expression with simplified arguments
    return reconstruct(expr; args=simplified_args, lower=lower_simplified, upper=upper_simplified)
end

File: packages/earthsci-toolkit/src/expression.ts
/**
 * Expression structural operations for the ESM format
 *
 * This module provides utilities for analyzing and manipulating mathematical
 * expressions in the ESM format AST.
 */

import type { Expression, ExpressionNode, Model } from './types.js'
import { isNumericLiteral, type NumericLiteral } from './numeric-literal.js'
import { evaluateExpression } from './codegen.js'

/**
 * Type alias for better readability. Widened to accept `NumericLiteral`
 * leaves (per discretization RFC §5.4.1) alongside plain JS numbers.
 */
export type Expr = Expression | NumericLiteral

/**
 * Extract all variable references from an expression
 * @param expr Expression to analyze
 * @returns Set of variable names referenced in the expression
 */
export function freeVariables(expr: Expr): Set<string> {
  const variables = new Set<string>()

  if (typeof expr === 'string') {
    variables.add(expr)
  } else if (typeof expr === 'number' || isNumericLiteral(expr)) {
    // Numeric literals contain no variables
    return variables
  } else if (typeof expr === 'object' && (expr as ExpressionNode).op) {
    // ExpressionNode - recursively analyze arguments
    for (const arg of (expr as ExpressionNode).args) {
      const childVars = freeVariables(arg as Expr)
      childVars.forEach(v => variables.add(v))
    }
  }

  return variables
}

/**
 * Extract free parameters from an expression within a model context
 * @param expr Expression to analyze
 * @param model Model context to determine parameter vs state variables
 * @returns Set of parameter names referenced in the expression
 */
export function freeParameters(expr: Expr, model: Model): Set<string> {
  const allVars = freeVariables(expr)
  const parameters = new Set<string>()

  for (const varName of allVars) {
    const variable = model.variables[varName]
    if (variable && variable.type === 'parameter') {
      parameters.add(varName)
    }
  }

  return parameters
}

/**
 * Check if an expression contains a specific variable
 * @param expr Expression to search
 * @param varName Variable name to look for
 * @returns True if the variable appears in the expression
 */
export function contains(expr: Expr, varName: string): boolean {
  if (typeof expr === 'string') {
    return expr === varName
  } else if (typeof expr === 'number' || isNumericLiteral(expr)) {
    return false
  } else if (typeof expr === 'object' && (expr as ExpressionNode).op) {
    // ExpressionNode - recursively check arguments
    return (expr as ExpressionNode).args.some(arg => contains(arg as Expr, varName))
  }

  return false
}

/**
 * Simplify an expression using basic algebraic rules
 * @param expr Expression to simplify
 * @returns Simplified expression
 */
export function simplify(expr: Expr): Expr {
  if (typeof expr === 'number' || typeof expr === 'string') {
    return expr
  }

  if (isNumericLiteral(expr)) {
    // NumericLiteral leaves carry a kind tag that simplify is not
    // responsible for folding. Canonical int/float folding lives in
    // canonicalize() per RFC §5.4. Return unchanged.
    return expr
  }

  if (typeof expr === 'object' && (expr as ExpressionNode).op) {
    // First simplify all arguments recursively
    const simplifiedArgs = expr.args.map(arg => simplify(arg))

    // Apply simplification rules based on operator
    switch (expr.op) {
      case '+':
        // Remove zeros: x + 0 -> x
        const nonZeroTerms = simplifiedArgs.filter(arg => arg !== 0)
        if (nonZeroTerms.length === 0) return 0
        if (nonZeroTerms.length === 1) return nonZeroTerms[0]

        // Separate constants and variables for partial constant folding
        const constants = nonZeroTerms.filter(arg => typeof arg === 'number') as number[]
        const variables = nonZeroTerms.filter(arg => typeof arg !== 'number')

        // If all terms are constants, return the sum
        if (variables.length === 0) {
          return constants.reduce((sum, val) => sum + val, 0)
        }

        // If there are constants to fold, combine them
        if (constants.length > 1) {
          const constantSum = constants.reduce((sum, val) => sum + val, 0)
          if (constantSum === 0) {
            // If constant sum is zero, just return variables
            return variables.length === 1 ? variables[0] : { ...expr, args: variables as [Expression, ...Expression[]] }
          } else {
            // Include the folded constant with variables
            const finalTerms = [...variables, constantSum]
            return { ...expr, args: finalTerms as [Expression, ...Expression[]] }
          }
        }

        return { ...expr, args: nonZeroTerms as [Expression, ...Expression[]] }

      case '*':
        // Zero multiplication: x * 0 -> 0
        if (simplifiedArgs.some(arg => arg === 0)) return 0

        // Remove ones: x * 1 -> x
        const nonOneFactors = simplifiedArgs.filter(arg => arg !== 1)
        if (nonOneFactors.length === 0) return 1
        if (nonOneFactors.length === 1) return nonOneFactors[0]

        // Separate constants and variables for partial constant folding
        const constantFactors = nonOneFactors.filter(arg => typeof arg === 'number') as number[]
        const variableFactors = nonOneFactors.filter(arg => typeof arg !== 'number')

        // If all factors are constants, return the product
        if (variableFactors.length === 0) {
          return constantFactors.reduce((prod, val) => prod * val, 1)
        }

        // If there are constants to fold, combine them
        if (constantFactors.length > 1) {
          const constantProd = constantFactors.reduce((prod, val) => prod * val, 1)
          if (constantProd === 0) {
            return 0
          } else if (constantProd === 1) {
            // If constant product is one, just return variables
            return variableFactors.length === 1 ? variableFactors[0] : { ...expr, args: variableFactors as [Expression, ...Expression[]] }
          } else {
            // Include the folded constant with variables
            const finalFactors = [...variableFactors, constantProd]
            return { ...expr, args: finalFactors as [Expression, ...Expression[]] }
          }
        }

        return { ...expr, args: nonOneFactors as [Expression, ...Expression[]] }

      case '-':
        if (simplifiedArgs.length === 1) {
          // Unary minus: -(-x) -> x would need deeper analysis
          if (typeof simplifiedArgs[0] === 'number') {
            return -simplifiedArgs[0]
          }
        } else if (simplifiedArgs.length === 2) {
          // Binary subtraction: x - 0 -> x
          if (simplifiedArgs[1] === 0) return simplifiedArgs[0]

          // Constant folding
          if (typeof simplifiedArgs[0] === 'number' && typeof simplifiedArgs[1] === 'number') {
            return simplifiedArgs[0] - simplifiedArgs[1]
          }
        }

        return { ...expr, args: simplifiedArgs as [Expression, ...Expression[]] }

      case '/':
        if (simplifiedArgs.length === 2) {
          // x / 1 -> x
          if (simplifiedArgs[1] === 1) return simplifiedArgs[0]

          // 0 / x -> 0 (assuming x != 0)
          if (simplifiedArgs[0] === 0) return 0

          // Constant folding
          if (typeof simplifiedArgs[0] === 'number' && typeof simplifiedArgs[1] === 'number') {
            if (simplifiedArgs[1] === 0) throw new Error('Division by zero')
            return simplifiedArgs[0] / simplifiedArgs[1]
          }
        }

        return { ...expr, args: simplifiedArgs as [Expression, ...Expression[]] }

      case '^':
        if (simplifiedArgs.length === 2) {
          // x^0 -> 1
          if (simplifiedArgs[1] === 0) return 1

          // x^1 -> x
          if (simplifiedArgs[1] === 1) return simplifiedArgs[0]

          // 0^x -> 0 (assuming x > 0)
          if (simplifiedArgs[0] === 0) return 0

          // 1^x -> 1
          if (simplifiedArgs[0] === 1) return 1

          // Constant folding
          if (typeof simplifiedArgs[0] === 'number' && typeof simplifiedArgs[1] === 'number') {
            return Math.pow(simplifiedArgs[0], simplifiedArgs[1])
          }
        }

        return { ...expr, args: simplifiedArgs as [Expression, ...Expression[]] }

      default:
        // For other operators, apply constant folding if all args are
        // numeric. Folding goes through the official codegen runner so
        // we share one dispatch table with the per-call evaluator.
        if (simplifiedArgs.every(arg => typeof arg === 'number')) {
          try {
            const tempBindings = new Map<string, number>()
            return evaluateExpression({ ...expr, args: simplifiedArgs as [Expression, ...Expression[]] }, tempBindings)
          } catch {
            return { ...expr, args: simplifiedArgs as [Expression, ...Expression[]] }
          }
        }

        return { ...expr, args: simplifiedArgs as [Expression, ...Expression[]] }
    }
  }

  return expr
}